Weird. I've given a try at removing the duplicate code from `_mapreduce()` with an innocent looking commit 2541655, but now the PR fails at bootstrap type with a strange error:
```
92.680954 REPL.jl
error during bootstrap:
LoadError(at "sysimg.jl" line 315: LoadError(at "REPL.jl" line 415: Base.MethodError(f=Base.#isless(), args=(-1, nothing), world=0x0000000000002ef7)))
rec_backtrace at /home/travis/build/JuliaLang/julia/src/stackwalk.c:84
record_backtrace at /home/travis/build/JuliaLang/julia/src/task.c:239
jl_throw at /home/travis/build/JuliaLang/julia/src/task.c:565
eval at /home/travis/build/JuliaLang/julia/src/interpreter.c:501
jl_interpret_toplevel_expr at /home/travis/build/JuliaLang/julia/src/interpreter.c:34
jl_toplevel_eval_flex at /home/travis/build/JuliaLang/julia/src/toplevel.c:477
jl_eval_module_expr at /home/travis/build/JuliaLang/julia/src/toplevel.c:203
jl_toplevel_eval_flex at /home/travis/build/JuliaLang/julia/src/toplevel.c:480
jl_parse_eval_all at /home/travis/build/JuliaLang/julia/src/ast.c:873
jl_load at /home/travis/build/JuliaLang/julia/src/toplevel.c:616
_include at ./sysimg.jl:241
unknown function (ip: 0x2b0ae9f35213)
jl_call_fptr_internal at /home/travis/build/JuliaLang/julia/src/julia_internal.h:326 [inlined]
jl_call_method_internal at /home/travis/build/JuliaLang/julia/src/julia_internal.h:345 [inlined]
jl_apply_generic at /home/travis/build/JuliaLang/julia/src/gf.c:2225
include at ./sysimg.jl:12
unknown function (ip: 0x2b0aea281c93)
jl_call_fptr_internal at /home/travis/build/JuliaLang/julia/src/julia_internal.h:326 [inlined]
jl_call_method_internal at /home/travis/build/JuliaLang/julia/src/julia_internal.h:345 [inlined]
jl_apply_generic at /home/travis/build/JuliaLang/julia/src/gf.c:2225
do_call at /home/travis/build/JuliaLang/julia/src/interpreter.c:75
eval at /home/travis/build/JuliaLang/julia/src/interpreter.c:242
jl_interpret_toplevel_expr at /home/travis/build/JuliaLang/julia/src/interpreter.c:34
jl_toplevel_eval_flex at /home/travis/build/JuliaLang/julia/src/toplevel.c:577
jl_eval_module_expr at /home/travis/build/JuliaLang/julia/src/toplevel.c:203
jl_toplevel_eval_flex at /home/travis/build/JuliaLang/julia/src/toplevel.c:480
jl_parse_eval_all at /home/travis/build/JuliaLang/julia/src/ast.c:873
jl_load at /home/travis/build/JuliaLang/julia/src/toplevel.c:616
exec_program at /home/travis/build/JuliaLang/julia/ui/repl.c:49
true_main at /home/travis/build/JuliaLang/julia/ui/repl.c:139
main at /home/travis/build/JuliaLang/julia/ui/repl.c:264
__libc_start_main at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
unknown function (ip: 0x4015f0)
92.963630make[2]: *** [/home/travis/build/JuliaLang/julia/usr/lib/julia/sys.o] Error 1 *** This error is usually fixed by running `make clean`. If the error persists, try `make cleanall`. ***
```

In AppVeyor the error looks like this:
```
error during bootstrap:
LoadError(at "sysimg.jl" line 315: LoadError(at "REPL.jl" line 415: Base.MethodError(f=Base.Terminals.jl
#isless(LineEdit.jl
REPLCompletions.jl
), REPL.jl
args=(-1, nothing), world=0x00002f0f)))
*** This error is usually fixed by running `make clean`. If the error persists, try `make cleanall`. ***
make[1]: *** [Makefile:228: /c/projects/julia/usr/lib/julia/sys.o] Error 1
make[1]: Leaving directory '/c/projects/julia'
make: *** [Makefile:105: julia-sysimg-release] Error 2
```

--- a/base/reduce.jl
+++ b/base/reduce.jl
@@ -168,7 +168,9 @@ foldr(op, itr) = mapfoldr(identity, op, itr)
 ## reduce & mapreduce
 
 function mapreduce_impl(f, op, A::AbstractArray, ifirst::Integer, ilast::Integer, blksize::Int=pairwise_blocksize(f, op))
-    if ifirst + blksize > ilast
+    if ifirst == ilast
+        return r_promote(op, f(A[ifirst]))
+    elseif ifirst + blksize > ilast
         # sequential portion
         fx1 = r_promote(op, f(A[ifirst]))
         fx2 = r_promote(op, f(A[ifirst + 1]))

--- a/test/reduce.jl
+++ b/test/reduce.jl
@@ -37,10 +37,14 @@
 @test mapreduce(-, +, [-10 -9 -3]) == ((10 + 9) + 3)
 @test mapreduce((x)->x[1:3], (x,y)->"($x+$y)", ["abcd", "efgh", "01234"]) == "((abc+efg)+012)"
 
-# mapreduce_impl() correctly works with 1- and n-sized blocks (PR #19325)
+# mapreduce_impl()
+#   correctly works with 1- and n-sized blocks (PR #19325)
 @test Base.mapreduce_impl(-, +, [-10, -9, -3], 2, 2) == 9
 @test Base.mapreduce_impl(abs2, +, [-10, -9, -3], 2, 3) == 81 + 9
 @test Base.mapreduce_impl(-, +, [-10, -9, -3, -4, -8, -2, -7], 2, 6, 2) == (9 + 3 + 4 + 8 + 2)
+#   type stability
+@test typeof(Base.mapreduce_impl(*, +, Int8[10], 1, 1)) === typeof(Base.mapreduce_impl(*, +, Int8[10, 11], 1, 2)) === typeof(Base.mapreduce_impl(*, +, Int8[10, 11, 12, 13], 1, 4))
+@test typeof(Base.mapreduce_impl(*, +, Float32[10.0], 1, 1)) === typeof(Base.mapreduce_impl(*, +, Float32[10, 11], 1, 2)) === typeof(Base.mapreduce_impl(*, +, Float32[10, 11, 12, 13], 1, 4))
 
 # sum
 

--- a/test/reduce.jl
+++ b/test/reduce.jl
@@ -43,8 +43,12 @@
 @test Base.mapreduce_impl(abs2, +, [-10, -9, -3], 2, 3) == 81 + 9
 @test Base.mapreduce_impl(-, +, [-10, -9, -3, -4, -8, -2, -7], 2, 6, 2) == (9 + 3 + 4 + 8 + 2)
 #   type stability
-@test typeof(Base.mapreduce_impl(*, +, Int8[10], 1, 1)) === typeof(Base.mapreduce_impl(*, +, Int8[10, 11], 1, 2)) === typeof(Base.mapreduce_impl(*, +, Int8[10, 11, 12, 13], 1, 4))
-@test typeof(Base.mapreduce_impl(*, +, Float32[10.0], 1, 1)) === typeof(Base.mapreduce_impl(*, +, Float32[10, 11], 1, 2)) === typeof(Base.mapreduce_impl(*, +, Float32[10, 11, 12, 13], 1, 4))
+@test typeof(Base.mapreduce_impl(*, +, Int8[10], 1, 1)) ===
+    typeof(Base.mapreduce_impl(*, +, Int8[10, 11], 1, 2)) ===
+    typeof(Base.mapreduce_impl(*, +, Int8[10, 11, 12, 13], 1, 4))
+@test typeof(Base.mapreduce_impl(*, +, Float32[10.0], 1, 1)) ===
+    typeof(Base.mapreduce_impl(*, +, Float32[10, 11], 1, 2)) ===
+    typeof(Base.mapreduce_impl(*, +, Float32[10, 11, 12, 13], 1, 4))
 
 # sum
 

--- a/test/reduce.jl
+++ b/test/reduce.jl
@@ -37,6 +37,11 @@
 @test mapreduce(-, +, [-10 -9 -3]) == ((10 + 9) + 3)
 @test mapreduce((x)->x[1:3], (x,y)->"($x+$y)", ["abcd", "efgh", "01234"]) == "((abc+efg)+012)"
 
+# mapreduce_impl() correctly works with 1- and n-sized blocks (PR #19325)
+@test Base.mapreduce_impl(-, +, [-10, -9, -3], 2, 2) == 9
+@test Base.mapreduce_impl(abs2, +, [-10, -9, -3], 2, 3) == 81 + 9
+@test Base.mapreduce_impl(-, +, [-10, -9, -3, -4, -8, -2, -7], 2, 6, 2) == (9 + 3 + 4 + 8 + 2)
+
 # sum
 
 @test sum(Int8[]) === Int32(0)

